### ISSUE TITLE
perf(gemma4): ~15–20% vLLM/server decode gap vs metal Direct (31B LB/QB2 apples-to-apples)

### ISSUE THREAD

**Issue by arginugaTT** [tenstorrent/tt-metal#51186]
## Summary

vLLM / tt-inference-server decode for Gemma4 is **~15–20% slower** than tt-metal Direct demo decode on the same Blackhole SKU when compared apples-to-apples (B=1 fair decode = `1000 / Mean_TPOT` or Peak output tok/s — **not** client Output token throughput, which folds in TTFT/queue).

Headline gap: **~19–20.5%** on `google/gemma-4-31B-it` @ P150x8 (LoudBox) for ISL 4k–32k.

Related tt-metal PR: https://github.com/tenstorrent/tt-metal/pull/50648

## Reproduction branches / SHAs

| Repo | Branch | Commit | Notes |
|------|--------|--------|-------|
| **tt-metal** | [`arg/gemma4_1x8optiom`](https://github.com/tenstorrent/tt-metal/tree/arg/gemma4_1x8optiom) | `054f10a3098` (`fix(gemma4): enable P150x8 vLLM 256k with multi-batch decode traces`, 2026-07-24) | SHA used for the Jul 24 metal↔server parity measurements |
| **tt-metal** (same branch tip) | [`arg/gemma4_1x8optiom`](https://github.com/tenstorrent/tt-metal/tree/arg/gemma4_1x8optiom) | `e60f866cf92` | Tip at ticket filing (later CI/PCC fixes on top of measure SHA); PR [#50648](https://github.com/tenstorrent/tt-metal/pull/50648) |
| **tenstorrent/vllm** | [`arg/gemma4_fixes`](https://github.com/tenstorrent/vllm/tree/arg/gemma4_fixes) | `06633f825` (`fix(tt-plugin): pad decode to nearest warmed batch bucket`) | Editable install used for server; parent/base `2593865d8` matches Gemma4 prod `vllm_commit: "2593865"` (token-chunked prefill / [#448](https://github.com/tenstorrent/vllm/pull/448)) |
| **tt-inference-server** | [`arg/gemma4_optimizations`](https://github.com/tenstorrent/tt-inference-server/tree/arg/gemma4_optimizations) | `17d50b3e9` (`Enable Gemma4-12B/31B at 256k on QB2 and LoudBox with policy chunked prefill`) | Local-server + benchmarks via committed model specs |

**Host:** Blackhole LoudBox. LB = `MESH_DEVICE=P150x8` (devices 0–7). QB2 = `P150x4` / `P300x2` via `TT_VISIBLE_DEVICES=4,5,6,7`.

## Methodology (apples-to-apples)

| Side | Path (committed) | Decode metric | Prefill metric |
|------|------------------|---------------|----------------|
| **Metal Direct** | `models/demos/gemma4/demo/text_demo.py` / `text_demo_v2.py`, B=1 | tok/s/user (steady decode) | TTFT (ms) |
| **Server (fair)** | `tt-inference-server` local server + benchmarks workflow (`run.py`), **max_concurrency=1** | `1000 / Mean_TPOT` (≈ Peak output tok/s) | Mean TTFT (ms) |
| **Server (unfair / do not use for decode gap)** | same | client **Output token throughput** | includes queue + TTFT → collapses at long ISL |

`gap_decode% = (metal_tok/s − server_1/TPOT) / metal_tok/s × 100`  
`gap_ttft% = (server_TTFT − metal_TTFT) / metal_TTFT × 100` (positive = server worse)

Optional device-side logs: `[gemma4-vllm-perf]` lines in committed `models/demos/gemma4/tt/generator_vllm.py` (useful for prefill/decode timing, but short-window decode samples are **not** the fair end-to-end decode metric).

---

## 1) Primary: Gemma-4-31B @ P150x8 (LoudBox) — ~20% decode gap

Metal Direct B=1 vs server B=1 fair decode (`1000/Mean_TPOT`).

| ISL | Metal tok/s | Server 1/TPOT | Server Peak | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | **gap_ttft%** | Server client out tok/s (unfair) |
|----:|------------:|--------------:|------------:|----------------:|----------------:|-----------------:|--------------:|---------------------------------:|
| 4096 | 32.46 | 26.29 | ~27 | **19.0%** | 658 | 740 | **12.5%** | 22.98 |
| 32768 | 31.60 | 25.13 | ~25 | **20.5%** | 6344 | 7359 | **16.0%** | 10.31 |
| 65536 | 29.61 | 24.50 | ~25 | **17.2%** | 14095 | 17311 | **22.8%** | 5.69 |
| 131072 | 26.15 | 22.33 | ~23 | **14.6%** | 28768 | 46256 | **60.8%** | 2.46 |
| 262144 | 21.00 | n/a | n/a | n/a | 126562 | n/a | n/a | n/a |

**Server B=1 short-ISL reference (same sweep):** ISL=128 → 1/TPOT=27.19, TTFT=89.2 ms, client tput=26.89.

### B=32 (throughput, not 1:1 with Direct B=1)

| Source | ISL | Metric | Value |
|--------|----:|--------|------:|
| Metal Direct | 4096 | tok/s/user / total | 21.4 / 684.7 |
| Server | 128 | client out / Mean TPOT | 436.7 / 55.04 ms (⇒ ~18.2 tok/s/user if 1/TPOT) |
| Server | 4096 | client out / TPOT | 130.6 / 153.08 ms |

---

## 2) Gemma-4-31B @ QB2 (P150x4 / P300x2) — same ~19–20% fair decode gap

SKU-matched metal Direct vs server B=1.

| ISL | Metal tok/s | Server 1/TPOT | Server Peak | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | **gap_ttft%** |
|----:|------------:|--------------:|------------:|----------------:|----------------:|-----------------:|--------------:|
| 4096 | 22.68 | 18.35 | 19 | **19.1%** | 1060 | 2558 | **141%** |
| 32768 | 22.74 | 18.09 | 19 | **20.4%** | 17120 | 24796 | **44.8%** |
| 65536 | 21.70 | 17.53 | 18 | **19.2%** | 27040 | 55148 | **103.9%** |

Note: do **not** compare metal LB Direct (~32 tok/s) to QB2 server (~18 tok/s) — that is a SKU mismatch.

---

## 3) Gemma-4-12B @ QB2

| ISL | Metal QB2 tok/s | Server 1/TPOT | Server Peak | **gap_decode%** (vs metal QB2) | Metal QB2 TTFT (ms) | Server TTFT (ms) | **gap_ttft%** |
|----:|----------------:|--------------:|------------:|-------------------------------:|--------------------:|-----------------:|--------------:|
| 4096 | 22.75 | 25.61 | 26 | **−12.6%** (server faster) | 1116 | 731 | **−34.5%** |
| 32768 | 22.20 | 24.84 | 25 | **−11.9%** (server faster) | 19268 | 12253 | **−36.4%** |
| 65536 | n/a | 24.93 | 25 | n/a | n/a | 26708 | n/a |
| 131072 | n/a | 22.87 | 23 | n/a | n/a | 62651 | n/a |

**12B LB Direct** (measured on same metal branch): 4k=47.59, 32k=44.44, 64k=41.32, 128k=37.51 tok/s — **no matching long-ISL B=1 server LB sweep** completed (12B LB bench hit EngineCore bus error after boot).

**12B QB2 short B=1:** ISL=128 → 1/TPOT=27.11, Peak=28, TTFT=115 ms.

---

## 4) Gemma-4-26B-A4B — incomplete apples-to-apples

| Side | Result |
|------|--------|
| Metal Direct | No stable long-ISL Direct scoreboard on LB/QB2 in this campaign (OOM/bus) |
| Server nightly-like (QB2) | PASS; Mean TPOT ≈ 68.2 ms ⇒ **~14.7 tok/s** fair decode; Mean TTFT ≈ 141 s (nightly-like concurrent short prompts — not B=1 ISL-matched) |

Cannot claim a clean % regression vs metal Direct until metal 26B Direct baselines exist on the same SKU/ISL.

---

## 5) Gemma-4-E2B / E4B

Metal Direct can be measured via the committed Gemma4 demos on this branch. **No vLLM server parity sweeps** were run for E2B/E4B in this campaign — N/A for metal-vs-server %.

---

## Device-side vs client (31B LB)

`[gemma4-vllm-perf]` decode samples in server logs often show **~140–170 tok/s/user** for short windows — that is **not** comparable to metal’s end-to-end demo tok/s (different accounting / partial windows). Fair end-to-end decode remains **Mean TPOT / Peak**.

Prefill: metal vs server TTFT at short ISL is closer; at 128k server TTFT is **~61% worse** than metal Direct on LB 31B (see table §1).

---

## Expected / ask

1. Close the **~15–20% B=1 fair decode** gap between metal Direct and vLLM server for **31B on P150x8 and QB2** (same topology, same ISL, Mean TPOT).
2. Investigate **TTFT** regression at long ISL (especially 64k/128k) on server vs metal Direct.
3. Finish apples-to-apples for **12B LB** (retry after bus error) and **26B** once metal Direct is stable.
4. Keep using **1/TPOT or Peak**, not client Output token throughput, for decode parity dashboards.

### Metal Direct (tt-metal `arg/gemma4_1x8optiom`)

```bash
cd tt-metal && source python_env/bin/activate
export HF_MODEL=google/gemma-4-31B-it MESH_DEVICE=P150x8
# Example long-context Direct (committed demo):
pytest models/demos/gemma4/demo/text_demo.py -k long-context-128k -s --timeout 1800
# Also: models/demos/gemma4/demo/text_demo_v2.py (same MESH_DEVICE / HF_MODEL pattern)
```

For QB2 Direct, use `MESH_DEVICE=P150x4` (or `P300x2`) with `TT_VISIBLE_DEVICES` set to the 4-chip subset.

### Server + benchmarks

1. Build/install **tt-metal** from `arg/gemma4_1x8optiom` (measure SHA `054f10a3098` or tip).
2. Install **tenstorrent/vllm** from branch `arg/gemma4_fixes` @ `06633f825` (or at least `2593865`) editable into that env, plus `plugins/vllm-tt-plugin`.
3. Use **tt-inference-server** branch `arg/gemma4_optimizations` @ `17d50b3e9` with local server + benchmarks against the committed Gemma4 specs in `workflows/model_specs/` (prod pins `vllm_commit: "2593865"` for gemma-4-12B/26B/31B; runtime used the `arg/gemma4_fixes` tip above).
4. Compare B=1 points using **Mean TPOT → 1000/TPOT** (and Mean TTFT) against metal Direct tok/s/user + TTFT at the same ISL/SKU.

Nightly-style vLLM harness on the metal side (committed): `.github/workflows/vllm-nightly-tests-impl.yaml` (defaults `vllm-commit: dev`; this campaign used `arg/gemma4_fixes` instead).

**Comment by viktorpusTT** [comment]
When tok/s is inverted to step duration, we get about 7 ms on LB and 11 on QB2, regardless of ISL. That looks like a constant per-step overhead.

The biggest suspect is the fact that Gemma4 doesn't support async scheduling: https://github.com/tenstorrent/tt-metal/blob/main/models/demos/gemma4/tt/generator_vllm.py#L118

That means vllm overhead is not hidden behind the previous/next step work. All is serialized, device is idle while the CPU is working.

To enable it, roughly the following must work:
* `model.decode_forward(..., read_from_device=False)`
* `model.read_decode_output(tt_out, async_read=True)`
* On-device sampling

Specifically for Gemma4 I have a feeling that we need to be pretty careful about warmup. If some traces are captured (and buffers reallocated for bigger size) after warmup, while vllm already runs decode and doesn't know about this realloc, that could lead to a big mess.

**Comment by arginugaTT** [comment]
## Update (2026-07-28): async decode enablement + **recalibrated** same-day B=1 ISL sweep

### Changes (same path for LB and QB2 — QB2 is not missing these)

**tt-metal** (`arg/gemma4_1x8optiom`, HEAD `5dad7d13ac7` + **uncommitted** Gemma4-local async continuity work):
- Declare `supports_async_decode` default **ON** for Gemma4 dense/MoE (kill-switch `GEMMA4_SUPPORTS_ASYNC_DECODE=0`). PLI (E2B/E4B) still forced **OFF**.
- Fix async continuity / token feedback lag under `async_scheduling` (**Gemma4 folder only** — no `tt_transformers` edits):
  - `models/demos/gemma4/tt/model.py`: non-PLI `always_refresh=False`; pad decode tokens to `[1,1,1,32]`; `ttnn.plus_one` on positions when `on_device_logits=True`
  - `models/demos/gemma4/tt/generator.py`: eager sampling writeback into feedback buffer via `ChunkedPrefillPageTableGuardMixin.sample_decode_on_device` (used by both Direct + vLLM generators)
  - `models/demos/gemma4/tt/generator_vllm.py`: async support + less decode sync when async ON
- Unit coverage: `models/demos/gemma4/tests/unit/test_async_decode_continuity.py`
- SKU-agnostic generator path → **QB2 P150x4 / P300x2 gets the same enablement + continuity fix as LoudBox P150x8**.

**tt-inference-server** (`arg/gemma4_optimizations`, HEAD `2c0a9b9d4` + **uncommitted** yaml):
- `async_scheduling: true` on **both** `P150X8` (LB) and `P300X2` (QB2) for `gemma-4-12B-it`, `gemma-4-31B-it`, and `gemma-4-26B-A4B-it` in `workflows/model_specs/{dev,prod}/llm.yaml`.
- vLLM: `arg/gemma4_fixes` @ `06633f825`.

### Recal methodology (2026-07-28T20:48Z UTC)
- Host: Blackhole LoudBox. LB = full `P150x8`. QB2 = dies `4,5,6,7`, `MESH_DEVICE=P150x4` / device `p300x2`.
- Fair decode = `1000 / Mean_TPOT` (B=1, OSL=128). `gap_decode% = (metal − server) / metal × 100` (negative = server ahead).
- **Both metal Direct and server measured in the same recal campaign** (`text_demo_v2.py` long-context-4k/32k/128k vs local server + benchmarks). Do **not** mix these metal numbers with the older issue-body scoreboard (LB Direct was higher there).
- **TTFT caveat:** not a guaranteed warmed-up protocol; directional only. Decode (`1/TPOT`) is the primary parity metric.
- Logs: `tt-metal/isl_sweep_logs/async_recal_20260728/` (`summary.tsv`).

### Recalibrated results (async ON)

#### 1) Gemma-4-31B @ P150x8 (LB) — **server ahead of same-day Direct (~−5% to −10%)**

| ISL | Metal tok/s | Server 1/TPOT | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | Coherence |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| 4096 | 27.63 | 30.29 | **−9.6%** | 699 | 779 | PASS |
| 32768 | 26.81 | 29.02 | **−8.2%** | 6781 | 7485 | PASS |
| 131072 | 23.95 | 25.18 | **−5.2%** | 31451 | 47297 | PASS |

vs original issue ~19–20% (server behind older metal scoreboard). Same-day Direct is lower than that scoreboard; with async ON, fair decode is **at/above** Direct.

#### 2) Gemma-4-31B @ QB2 (P150x4) — **gap ~13–15%** (improved vs ~19–20% in issue body)

| ISL | Metal tok/s | Server 1/TPOT | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | Coherence |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| 4096 | 23.33 | 19.94 | **14.5%** | 1059 | 1134 | PASS |
| 32768 | 22.22 | 19.36 | **12.9%** | 17754 | 11829 | PASS |
| 131072 | 20.43 | 17.58 | **14.0%** | 57220 | 82311 | PASS |

#### 3) Gemma-4-12B @ P150x8 (LB) — **server ahead (~−8% to −15%); long-coherence FAIL**

| ISL | Metal tok/s | Server 1/TPOT | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | Coherence |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| 4096 | 42.40 | 48.73 | **−14.9%** | 457 | 439 | FAIL (long) |
| 32768 | 41.80 | 46.30 | **−10.8%** | 6075 | 4019 | FAIL (long) |
| 131072 | 35.77 | 38.77 | **−8.4%** | 20920 | 22874 | FAIL (long) |

Short coherence OK; long padded prompt produced garbage (`… a... a...`). Decode numbers still from the B=1 bench sweep. **Do not treat LB 12B as quality-closed.**

#### 4) Gemma-4-12B @ QB2 — **gap ~17–19%** (same async path; coherence PASS)

| ISL | Metal tok/s | Server 1/TPOT | **gap_decode%** | Metal TTFT (ms) | Server TTFT (ms) | Coherence |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| 4096 | 34.71 | 28.71 | **17.3%** | 693 | 629 | PASS |
| 32768 | 34.47 | 27.89 | **19.1%** | 6480 | 5972 | PASS |
| 131072 | 30.67 | 24.95 | **18.7%** | 27824 | 37619 | PASS |

#### 5) Gemma-4-26B-A4B @ P150x8 (prior async matrix; not in this recal)

| ISL | Server 1/TPOT | Server TTFT (ms) |
| --- | ---: | ---: |
| 4096 | 19.65 | 65198 |
| 32768 | 19.45 | 521968 |
| 65536 | 19.17 | 1045173 |

Still no clean metal Direct scoreboard for % gap.

### Status / remaining ask
- **LB 31B:** original ~19–20% → **same-day gap −5% to −10%** (server ahead of Direct). Primary remaining work is metal Direct uplift / understanding why Direct is below the older scoreboard, not closing a server deficit.
- **QB2 31B:** original ~19–20% → **~13–15%** with the **same** async enablement.
- **QB2 12B:** still **~17–19%** vs same-day Direct.
- **LB 12B:** decode looks ahead of Direct, but **long-context coherence FAIL** must be fixed before calling parity done.
- Async metal (Gemma4-only) + inference-server yaml still **local/uncommitted**.
- TTFT: use with caution until a warmed-up TTFT protocol is agreed.
